src/NuGet.Core/NuGet.Protocol/CachingSourceProvider.cs

```diff
@@ -24,18 +24,29 @@ public class CachingSourceProvider : ISourceRepositoryProvider
         private readonly ConcurrentDictionary<string, SourceRepository> _cachedSources
             = new ConcurrentDictionary<string, SourceRepository>(StringComparer.Ordinal);
 
-        public CachingSourceProvider(IPackageSourceProvider packageSourceProvider)
+        public CachingSourceProvider(IPackageSourceProvider packageSourceProvider,
+            IEnumerable<Lazy<INuGetResourceProvider>> resourceProviders)
```

Comment:
I think it is better if we pass `custom implementations of resource providers` to this parameterized constructor instead of expecting the caller to always concat out of the resource providers and custom ones. The public constructor that that accepts only one parameter `IPackageSourceProvider` can invoke this constructor with an empty collection of custom providers. For e.g. the new implementation may look like below.

```cs
        public CachingSourceProvider(IPackageSourceProvider packageSourceProvider,
            IEnumerable<Lazy<INuGetResourceProvider>> customResourceProviders)
        {
//null checking
            _packageSourceProvider = packageSourceProvider;

            _resourceProviders.AddRange(Repository.Provider.GetCoreV3());
            _resourceProviders.AddRange(customResourceProviders);
//code removed to focus on the proposed changes
        }

        public CachingSourceProvider(IPackageSourceProvider packageSourceProvider)
            : this(packageSourceProvider, Enumerable.Empty<Lazy<INuGetResourceProvider>>())
        {
        }
```

Reply:
@kartheekp-ms actually, in my use case, I override one of the providers from `Provider.GetCoreV3()` with my custom implementation. So, the original intention was to customize the default list of providers.

Reply:
I think customizing the default list of providers is already supported. The constructor of `ResourceProvider` accepts kind of precedence parameter used to decide which provider to be called first. https://github.com/NuGet/NuGet.Client/blob/4dbf16e68eabf507b9f156c801b1b7c3b0586e43/src/NuGet.Core/NuGet.Protocol/ResourceProvider.cs#L29-L31
Most of the default resource providers pass [`NuGetResourceProviderPositions.Last`](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Protocol/NuGetResourceProviderPositions.cs) as argument that is later used in [`sorting`](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Protocol/ProviderComparer.cs#L61-L83) logic.
https://github.com/NuGet/NuGet.Client/blob/4dbf16e68eabf507b9f156c801b1b7c3b0586e43/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpHandlerResourceV3Provider.cs#L17-L20

Out of curiosity, I tried to find any open source implementations making use of this feature and found a reference here where base constructor was invoked by passing `NuGetResourceProviderPositions.First` as argument.

https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/blob/3dde15714b4c76d28d4db2af59f43188ad83ed68/PackageExplorer/MefServices/PackageDownloader.cs#L331-L338

Reply:
@AndreyAkinshin - Please let us know if you need further clarifications on the above comment considering your use case.

Reply:
@AndreyAkinshin - As per [contributing guidelines](https://github.com/NuGet/NuGet.Client/blob/dev/CONTRIBUTING.md#contributing-step-by-step), `If we don't receive any feedback within 2 months from today, the PR will be closed due to inactivity`. Please let me know If you need additional information regarding the feedback provided in this PR.

Reply:
@kartheekp-ms OK, it seems that an additional `CachingSourceProvider` constructor with `customResourceProviders` should solve my problem. 
Should I create another PR with the implementation of the second constructor?

Reply:
@kartheekp-ms is out of the office at the moment, so I'll take over for now. I apologise for not chiming in earlier.

I like Kartheek's suggestion of having it pass in additional resource providers, instead of the full list of providers. I think it makes the API to be used more easily than needing to concat custom resource providers. However, there are other APIs which already expect the list of resource providers to be the full list, not an additional list. So I think it would be confusing that some APIs behave in one way and other APIs work in another way. Additionally, with the way that .NET works, it's impossible to have:

```c#
public CachingSourceProvider(IPackageSourceProvider packageSourceProvider, IEnumerable<Lazy<INuGetResourceProvider>> allResourceProviders)
public CachingSourceProvider(IPackageSourceProvider packageSourceProvider, IEnumerable<Lazy<INuGetResourceProvider>> additionalResourceProviders)
```

as both methods have the same signature.

Therefore, if we're going to add a new constructor, I'd prefer it to keep the behaviour of the API already in this PR.

However, some time ago we added the following to support VS for Mac to provide their own ResourceProvider factory in a way that all NuGet APIs will use automatically:

https://github.com/NuGet/NuGet.Client/blob/fcc6c115ab6c01b6a0a6e09c77dde40b660cb00c/src/NuGet.Core/NuGet.Protocol/Repository.cs#L18-L33

@AndreyAkinshin I believe you can use this existing extensibility, rather than the one proposed by this PR. If you agree, please close this PR.  If you still think this PR's addition is valuable, please rebase your PR, as we are now using Roslyn's PublicApiAnalyzers, and you need to modify PublicApi.Unshipped.txt, which the analyzer will give you a code suggestion for once you rebase and reload the solution.

Reply:
@zivkan it seems that this customization point should resolve my problem. Thank you very much. I close this PR.